instrumentor/sources_webhooks.go

```diff
@@ -0,0 +1,263 @@
+		!controllerutil.ContainsFinalizer(source, consts.DeleteInstrumentationConfigFinalizer) {
+		controllerutil.AddFinalizer(source, consts.DeleteInstrumentationConfigFinalizer)
+	}
+	if v1alpha1.IsWorkloadExcludedSource(source) &&
+		source.DeletionTimestamp.IsZero() &&
+		!controllerutil.ContainsFinalizer(source, consts.StartLangDetectionFinalizer) {
+		controllerutil.AddFinalizer(source, consts.StartLangDetectionFinalizer)
+	}
+
+	return nil
+}
+
+type SourcesValidator struct {
+	client.Client
+}
+
+var _ webhook.CustomValidator = &SourcesValidator{}
+
+func (s *SourcesValidator) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	var allErrs field.ErrorList
+	source, ok := obj.(*v1alpha1.Source)
+	if !ok {
+		return nil, fmt.Errorf("expected a Source but got a %T", obj)
+	}
+
+	errs := s.validateSourceFields(ctx, source)
+	if len(errs) > 0 {
+		allErrs = append(allErrs, errs...)
+	}
+
+	if len(allErrs) == 0 {
+		return nil, nil
+	}
+
+	return nil, apierrors.NewInvalid(
+		schema.GroupKind{Group: "odigos.io", Kind: "Source"},
+		source.Name, allErrs)
+}
+
+func (s *SourcesValidator) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
+	var allErrs field.ErrorList
+	old, ok := oldObj.(*v1alpha1.Source)
+	if !ok {
+		return nil, fmt.Errorf("expected a Source but got a %T", old)
+	}
+	new, ok := newObj.(*v1alpha1.Source)
+	if !ok {
+		return nil, fmt.Errorf("expected a Source but got a %T", new)
+	}
+
+	if new.GetName() != old.GetName() {
+		allErrs = append(allErrs, field.Invalid(
+			field.NewPath("metadata").Child("name"),
+			new.GetName(),
+			"Source name is immutable",
+		))
+	}
+
+	if new.GetNamespace() != old.GetNamespace() {
+		allErrs = append(allErrs, field.Invalid(
+			field.NewPath("metadata").Child("namespace"),
+			new.GetNamespace(),
+			"Source namespace is immutable",
+		))
+	}
+
+	if new.Labels[consts.WorkloadKindLabel] != old.Labels[consts.WorkloadKindLabel] {
+		allErrs = append(allErrs, field.Invalid(
+			field.NewPath("metadata").Child("labels"),
+			new.Labels[consts.WorkloadKindLabel],
+			"Source workload-kind label is immutable",
+		))
+	}
+	if new.Labels[consts.WorkloadNameLabel] != old.Labels[consts.WorkloadNameLabel] {
+		allErrs = append(allErrs, field.Invalid(
+			field.NewPath("metadata").Child("labels"),
+			new.Labels[consts.WorkloadNameLabel],
+			"Source workload-name label is immutable",
+		))
+	}
+	if new.Labels[consts.WorkloadNamespaceLabel] != old.Labels[consts.WorkloadNamespaceLabel] {
+		allErrs = append(allErrs, field.Invalid(
+			field.NewPath("metadata").Child("labels"),
+			new.Labels[consts.WorkloadNamespaceLabel],
+			"Source workload-namespace label is immutable",
+		))
+	}
+	if new.Spec.Workload != old.Spec.Workload {
+		allErrs = append(allErrs, field.Invalid(
+			field.NewPath("spec").Child("workload"),
+			new.Spec.Workload,
+			"Source workload is immutable",
+		))
+	}
+
+	errs := s.validateSourceFields(ctx, new)
```

Comment:
If it's an update, and we've already verified that the labels and spec workload is not changed, what value do we get from calling it here? or is it only for safety?

Reply:
Yeah I figured it doesn't hurt to double check that they are also valid